File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.go
// Abstract Factory pattern in Go.
//
// Small per-product interfaces plus a factory interface; concrete factories
// are zero-size structs. Implicit interface satisfaction means no family
// declares that it implements anything.

package main

import "fmt"

type Button interface {
	Render() string
}

type Checkbox interface {
	Render() string
}

type macButton struct{}
type macCheckbox struct{}
type windowsButton struct{}
type windowsCheckbox struct{}

func (macButton) Render() string       { return "( OK )  // rounded macOS button" }
func (macCheckbox) Render() string     { return "◉ Remember me  // macOS checkbox" }
func (windowsButton) Render() string   { return "[ OK ]  // rectangular Windows button" }
func (windowsCheckbox) Render() string { return "[x] Remember me  // Windows checkbox" }

// WidgetFactory has one creation method per product kind; products are
// guaranteed to match.
type WidgetFactory interface {
	CreateButton() Button
	CreateCheckbox() Checkbox
}

type MacFactory struct{}

func (MacFactory) CreateButton() Button     { return macButton{} }
func (MacFactory) CreateCheckbox() Checkbox { return macCheckbox{} }

type WindowsFactory struct{}

func (WindowsFactory) CreateButton() Button     { return windowsButton{} }
func (WindowsFactory) CreateCheckbox() Checkbox { return windowsCheckbox{} }

// renderDialog is the client: it depends only on the factory interface.
func renderDialog(factory WidgetFactory) {
	fmt.Println(factory.CreateButton().Render())
	fmt.Println(factory.CreateCheckbox().Render())
}

func main() {
	themes := []struct {
		name    string
		factory WidgetFactory
	}{
		{"macOS", MacFactory{}},
		{"Windows", WindowsFactory{}},
	}

	for _, theme := range themes {
		fmt.Printf("-- %s theme --\n", theme.name)
		renderDialog(theme.factory)
	}
}

File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.java
// Abstract Factory pattern in Java.
//
// Java identifiers cannot contain hyphens, so unlike the other examples the
// class cannot be named after this file; a package-private class keeps the
// single-file source launcher (`java abstract-factory.java`) working.

class AbstractFactoryDemo {

    interface Button {
        String render();
    }

    interface Checkbox {
        String render();
    }

    static final class MacButton implements Button {
        public String render() { return "( OK )  // rounded macOS button"; }
    }

    static final class MacCheckbox implements Checkbox {
        public String render() { return "◉ Remember me  // macOS checkbox"; }
    }

    static final class WindowsButton implements Button {
        public String render() { return "[ OK ]  // rectangular Windows button"; }
    }

    static final class WindowsCheckbox implements Checkbox {
        public String render() { return "[x] Remember me  // Windows checkbox"; }
    }

    /** One creation method per product kind; products are guaranteed to match. */
    interface WidgetFactory {
        Button createButton();
        Checkbox createCheckbox();
    }

    static final class MacFactory implements WidgetFactory {
        public Button createButton() { return new MacButton(); }
        public Checkbox createCheckbox() { return new MacCheckbox(); }
    }

    static final class WindowsFactory implements WidgetFactory {
        public Button createButton() { return new WindowsButton(); }
        public Checkbox createCheckbox() { return new WindowsCheckbox(); }
    }

    /** Client code: depends only on the abstract factory and products. */
    static void renderDialog(WidgetFactory factory) {
        System.out.println(factory.createButton().render());
        System.out.println(factory.createCheckbox().render());
    }

    public static void main(String[] args) {
        System.out.println("-- macOS theme --");
        renderDialog(new MacFactory());

        System.out.println("-- Windows theme --");
        renderDialog(new WindowsFactory());
    }
}
